File: python/mirage/commands/builtin/langfuse/rg.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.langfuse import LangfuseAccessor
from mirage.commands.builtin.generic.rg import rg as generic_rg
from mirage.commands.builtin.generic_bind.adapter import bound_op
from mirage.commands.builtin.grep_helper import (compile_pattern,
                                                 has_search_shaping_flags,
                                                 pattern_arg)
from mirage.commands.builtin.langfuse.grep import (_filter_traces,
                                                   _format_dataset_results,
                                                   _format_prompt_results,
                                                   _format_session_results)
from mirage.commands.builtin.langfuse.io import resolve_glob
from mirage.commands.builtin.utils.paths import has_unresolved_glob
from mirage.commands.config import CommandOpts
from mirage.commands.errors import UsageError
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import FlagView
from mirage.core.langfuse.client import (fetch_datasets, fetch_prompts,
                                         fetch_sessions, fetch_traces)
from mirage.core.langfuse.read import read as langfuse_read
from mirage.core.langfuse.readdir import readdir as _readdir
from mirage.core.langfuse.scope import SEARCH_KINDS, detect_scope
from mirage.core.langfuse.stat import stat as _stat
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("rg", resource="langfuse", spec=SPECS["rg"])
async def rg(accessor: LangfuseAccessor, paths: list[PathSpec],
             texts: list[str],
             opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(opts.flags, spec=SPECS["rg"])
    pattern_str = pattern_arg(texts, fl)
    if pattern_str is None:
        raise UsageError("rg: usage: rg [flags] pattern [path]")
    i = fl.as_bool("i")
    w = fl.as_bool("w")
    F = fl.as_bool("F")
    pat = compile_pattern(pattern_str, i, F, w)

    config = accessor.config
    limit = config.default_search_limit

    # The search push-down answers from the list endpoints (one call instead
    # of one read per entry), so it greps listing summaries: a pattern that
    # only occurs in a trace's observation bodies needs a file read to match.
    # Output/match-shaping flags defer to the generic scan below.
    if (paths and not has_unresolved_glob(paths) and "\n" not in pattern_str
            and not has_search_shaping_flags(opts.flags)):
        search = SEARCH_KINDS.get(detect_scope(paths[0]).kind)

        if search == "traces":
            traces = await fetch_traces(
                accessor.api,
                limit=limit,
            )
            return _filter_traces(traces, pat)

        if search == "sessions":
            sessions = await fetch_sessions(
                accessor.api,
                limit=limit,
            )
            return _format_session_results(sessions, pat)

        if search == "prompts":
            prompts = await fetch_prompts(accessor.api)
            return _format_prompt_results(prompts, pat)

        if search == "datasets":
            datasets = await fetch_datasets(accessor.api)
            return _format_dataset_results(datasets, pat)

    resolved = await resolve_glob(accessor, paths,
                                  index=opts.index) if paths else []
    return await generic_rg(
        resolved,
        texts,
        opts.flags,
        readdir=bound_op(_readdir, accessor, opts.index),
        stat=bound_op(_stat, accessor, opts.index),
        read_bytes=bound_op(langfuse_read, accessor, opts.index),
        read_stream=None,
        stdin=opts.stdin,
    )

File: typescript/packages/core/src/commands/builtin/langfuse/rg.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../../accessor/langfuse.ts'
import type { IndexCacheStore } from '../../../cache/index/index.ts'
import {
  fetchDatasets,
  fetchPrompts,
  fetchSessions,
  fetchTraces,
} from '../../../core/langfuse/client.ts'
import { resolveGlobOf } from '../generic_bind/index.ts'
import { LANGFUSE_IO } from './io.ts'
import { read as langfuseRead } from '../../../core/langfuse/read.ts'
import { readdir as langfuseReaddir } from '../../../core/langfuse/readdir.ts'
import { SEARCH_KINDS, detectScope } from '../../../core/langfuse/scope.ts'
import { stat as langfuseStat } from '../../../core/langfuse/stat.ts'
import { type FileStat, ResourceName, type PathSpec } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { rgGeneric } from '../generic/rg.ts'
import { compilePattern, patternArg } from '../grep_helper.ts'
import { filterDatasets, filterPrompts, filterSessions, filterTraces } from './grep.ts'
import { FlagView } from '../../spec/types.ts'

const resolveLangfuseGlob = resolveGlobOf(LANGFUSE_IO)

async function* langfuseStream(
  accessor: LangfuseAccessor,
  p: PathSpec,
  index: IndexCacheStore | undefined,
): AsyncIterable<Uint8Array> {
  yield await langfuseRead(accessor, p, index)
}

async function rgCommand(
  accessor: LangfuseAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const pattern = patternArg(texts, opts.flags)
  const limit = accessor.config.defaultSearchLimit ?? 50

  const first = paths[0]
  if (first !== undefined && pattern !== null && !pattern.includes('\n')) {
    const search = SEARCH_KINDS[detectScope(first).kind]
    const fl = new FlagView(opts.flags, specOf('rg'))
    const pat = compilePattern(pattern, fl.asBool('i'), fl.asBool('F'), fl.asBool('w'))
    if (search === 'traces') {
      const traces = await fetchTraces(accessor.transport, { limit })
      return filterTraces(traces, pat)
    }
    if (search === 'sessions') {
      const sessions = await fetchSessions(accessor.transport, { limit })
      return filterSessions(sessions, pat)
    }
    if (search === 'prompts') {
      const prompts = await fetchPrompts(accessor.transport)
      return filterPrompts(prompts, pat)
    }
    if (search === 'datasets') {
      const datasets = await fetchDatasets(accessor.transport)
      return filterDatasets(datasets, pat)
    }
  }

  const resolved =
    paths.length > 0 ? await resolveLangfuseGlob(accessor, paths, opts.index ?? undefined) : []
  const stat = (p: PathSpec): Promise<FileStat> =>
    langfuseStat(accessor, p, opts.index ?? undefined)
  const readdir = (p: PathSpec): Promise<string[]> =>
    langfuseReaddir(accessor, p, opts.index ?? undefined)
  const stream = (p: PathSpec): AsyncIterable<Uint8Array> =>
    langfuseStream(accessor, p, opts.index ?? undefined)
  return rgGeneric(resolved, texts, opts, stat, readdir, stream)
}

export const LANGFUSE_RG = command({
  name: 'rg',
  resource: ResourceName.LANGFUSE,
  spec: specOf('rg'),
  fn: rgCommand,
})
